### PR TITLE
Remove circular deps on gcc when using clang

### PR DESCRIPTION
**Pull request by lyssdod**
This change allows to have cleanly linked `libcxx`, `libcxxabi` and `libunwind` libs when using Clang as a main system compiler:

```
$ ldd /usr/lib/libc++abi.so 
	ldd (0x7f9741fbc000)
	libunwind.so.1 => /usr/lib/libunwind.so.1 (0x7f9741f6e000)
	libc.so => ldd (0x7f9741fbc000)
```

```
$ ldd /usr/lib/libunwind.so 
	ldd (0x7f18a9723000)
	libc.so => ldd (0x7f18a9723000)
```

Default configuration mentioned on the [wiki](https://wiki.gentoo.org/wiki/Clang#Bootstrapping_the_Clang_toolchain) results in these libs linking either to `libgcc_s`, `libatomic`, or both. Final USE flags are:

```
sys-devel/clang compiler-rt default-compiler-rt default-lld default-libcxx llvm-libunwind
sys-devel/clang-runtime compiler-rt libcxx
sys-libs/compiler-rt clang
sys-libs/compiler-rt-sanitizers clang
sys-libs/libcxx libunwind libcxxabi
sys-libs/libcxxabi libunwind
sys-libs/llvm-libunwind compiler-rt
```

### CODE REVIEW

File: sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild [hunk 1]
```diff
@@ -13,7 +13,7 @@ HOMEPAGE="https://github.com/llvm-mirror/libunwind"
 LICENSE="Apache-2.0-with-LLVM-exceptions || ( UoI-NCSA MIT )"
 SLOT="0"
 KEYWORDS="amd64 arm arm64 ~ppc64 ~riscv x86 ~x64-macos"
-IUSE="debug static-libs test"
+IUSE="debug static-libs test compiler-rt"
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild:16]: I don't think this needs to be a USE flag. If you really need to switch the CMake flag explicitly, it's better to detect what rtlib clang uses. I think one of the packages has code for that that you can reuse.

**Review by mgorny — CHANGES_REQUESTED** [review]
1. Please take look at other commit messages, and fix yours to conform.
2. You need to use your full name in signoff.

File: sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild (hunk 1 above)
Comment (lyssdod) [reply sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild:16]: That's the trick...both `clang` and `clang-runtime` have this USE flag, but there's no autodetection. Another option is to switch it ON permanently – but then `compiler-rt` needs to be a hard dependency, and that doesn't cover all possible use cases and seems like against Gentoo spirit to me.

**Comment by lyssdod** [comment]
Formatted commit messages

File: sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild (hunk 1 above)
Comment (lyssdod) [reply sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild:16]: I think I got what you mean, looks like a job for `has_use` but I cannot find a proper example

Comment (lyssdod) [reply sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild]: This should do

File: sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild [hunk 2]
```diff
@@ -23,7 +23,8 @@ DEPEND="
 BDEPEND="
 	test? ( >=sys-devel/clang-3.9.0
 		$(python_gen_any_dep 'dev-python/lit[${PYTHON_USEDEP}]')
-	)"
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild:26]: Don't change the formatting here.

File: sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild [hunk 3]
```diff
@@ -38,8 +39,19 @@ pkg_setup() {
 }
 
 multilib_src_configure() {
+	local crtstr=""
 	local libdir=$(get_libdir)
 
+	if ! has_version "sys-libs/compiler-rt"; then
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild:45]: Nope, not that. Look at libcxx ebuilds.

File: sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild (hunk 2 above)
Comment (lyssdod) [reply sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild]: Fixed

File: sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild (hunk 3 above)
Comment (lyssdod) [reply sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild]: Fixed

File: sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild [hunk 4]
```diff
@@ -38,8 +38,18 @@ pkg_setup() {
 }
 
 multilib_src_configure() {
+	local use_compiler_rt=OFF
 	local libdir=$(get_libdir)
 
+	# link to compiler-rt if available
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild:44]: 'if available' is no longer correct.

**Review by mgorny — COMMENTED** [review]
Could you please squash your commits now?

**Comment by lyssdod** [comment]
> Could you please squash your commits now?

Sure, but should I put an arbitrary (e.g. like PR's title) commit message, since this affects 2 packages?

**Comment by mgorny** [comment]
One commit per package, please.

File: sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild [hunk 5]
```diff
@@ -38,8 +38,18 @@ pkg_setup() {
 }
 
 multilib_src_configure() {
+	local use_compiler_rt=OFF
 	local libdir=$(get_libdir)
 
+	# link to compiler-rt
+	if tc-is-clang; then
```
Comment (mgorny) [inline sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild:45]: You need to inherit toolchain-funcs for this function.

**Comment by mgorny** [comment]
You seem to have the changes mixed up between the two commits.

**Comment by lyssdod** [comment]
Commits fixed

File: sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild [hunk 6]
```diff
@@ -38,8 +38,18 @@ pkg_setup() {
 }
 
 multilib_src_configure() {
+	local use_compiler_rt=OFF
 	local libdir=$(get_libdir)
 
+	# link to compiler-rt
```
Comment (thesamesam) [inline sys-libs/llvm-libunwind/llvm-libunwind-12.0.0.ebuild:44]: I'd consider adding a link to this PR so we don't forget why we did it.

**Review by mgorny — APPROVED** [review]
Thank you. I'm sorry it took so many rounds. I'll do a final fixup and merge it.

**Comment by lyssdod** [comment]
Thank you!

**Comment by 12101111** [comment]
You misspell `compiler_rt` into `compiler-rt`:

```
>>> Configuring source in /tmp/portage/sys-libs/llvm-libunwind-12.0.1/work/libunwind ...
 * abi_x86_64.amd64: running multilib-minimal_abi_src_configure
clang-12: warning: argument unused during compilation: '-E' [-Wunused-command-line-argument]
/tmp/portage/sys-libs/llvm-libunwind-12.0.1/temp/environment: line 4234: local: `compiler-rt=/usr/lib/llvm/12/bin/../../../../lib/clang/12.0.1/lib/linux/libclang_rt.builtins-x86_64.a': not a valid identifier
 * Working in BUILD_DIR: "/tmp/portage/sys-libs/llvm-libunwind-12.0.1/work/llvm-libunwind-12.0.1_build-abi_x86_64.amd64"
cmake -C /tmp/portage/sys-libs/llvm-libunwind-12.0.1/work/llvm-libunwind-12.0.1_build-abi_x86_64.amd64/gentoo_common_config.cmake -G Ninja -DCMAKE_INSTALL_PREFIX=/usr -DLLVM_LIBDIR_SUFFIX= -DLIBUNWIND_ENABLE_ASSERTIONS=no -DLIBUNWIND_ENABLE_STATIC=yes -DLLVM_INCLUDE_TESTS=no -DLIBUNWIND_ENABLE_CROSS_UNWINDING=ON -DLIBUNWIND_USE_COMPILER_RT=OFF -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_TOOLCHAIN_FILE=/tmp/portage/sys-libs/llvm-libunwind-12.0.1/work/llvm-libunwind-12.0.1_build-abi_x86_64.amd64/gentoo_toolchain.cmake  /tmp/portage/sys-libs/llvm-libunwind-12.0.1/work/libunwind
loading initial cache file /tmp/portage/sys-libs/llvm-libunwind-12.0.1/work/llvm-libunwind-12.0.1_build-abi_x86_64.amd64/gentoo_common_config.cmake
```

info bash say:

```
'name'
     A 'word' consisting solely of letters, numbers, and underscores,
     and beginning with a letter or underscore.  'Name's are used as
     shell variable and function names.  Also referred to as an
     'identifier'.
```

**Comment by thesamesam** [comment]
@12101111 In general, comments on a closed PR are easy to miss, so it's usually best to open a new bug. Anyway, I'll pass this on.

**Comment by mgorny** [comment]
Heh, for some reason I've fixed it in 12.0.1 but not in newer versions.

**Pull request closed without merging**